File: Delphi/ImapClient/MessageFrm.pas
unit MessageFrm;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, clMailMessage;

type
  TMessageForm = class(TForm)
    Label1: TLabel;
    Label2: TLabel;
    Label3: TLabel;
    edtFrom: TEdit;
    edtTo: TEdit;
    edtSubject: TEdit;
    memBody: TMemo;
    btnAppend: TButton;
    btnCancel: TButton;
  private
    { Private declarations }
  public
    class procedure ShowMessage(AMsg: TclMailMessage);
    class function ComposeMessage(AMsg: TclMailMessage): Boolean;
    { Public declarations }
  end;

var
  MessageForm: TMessageForm;

implementation

{$R *.dfm}

{ TMessageForm }

class function TMessageForm.ComposeMessage(AMsg: TclMailMessage): Boolean;
var
  dlg: TMessageForm;
begin
  dlg := TMessageForm.Create(nil);
  try
    dlg.btnAppend.Visible := True;
    dlg.btnCancel.Caption := 'Cancel';

    dlg.edtFrom.ReadOnly := False;
    dlg.edtTo.ReadOnly := False;
    dlg.edtSubject.ReadOnly := False;
    dlg.memBody.ReadOnly := False;

    dlg.edtFrom.Text := 'clevertester@company.mail';
    dlg.edtTo.Text := 'clevertester@company.mail';
    dlg.edtSubject.Text := 'Test Subject';
    dlg.memBody.Text := 'Test Body';

    Result := (dlg.ShowModal() = mrOk);

    if Result then
    begin
      AMsg.BuildMessage(dlg.memBody.Lines.Text, '');
      AMsg.From.FullAddress := dlg.edtFrom.Text;
      AMsg.ToList.EmailAddresses := dlg.edtTo.Text;
      AMsg.Subject := dlg.edtSubject.Text;
    end;
  finally
    dlg.Free();
  end;
end;

class procedure TMessageForm.ShowMessage(AMsg: TclMailMessage);
var
  dlg: TMessageForm;
begin
  dlg := TMessageForm.Create(nil);
  try
    dlg.btnAppend.Visible := False;
    dlg.btnCancel.Caption := 'Close';

    dlg.edtFrom.ReadOnly := True;
    dlg.edtTo.ReadOnly := True;
    dlg.edtSubject.ReadOnly := True;
    dlg.memBody.ReadOnly := True;

    dlg.edtFrom.Color := clWindow;
    dlg.edtTo.Color := clWindow;
    dlg.edtSubject.Color := clWindow;
    dlg.memBody.Color := clWindow;

    dlg.edtFrom.Text := AMsg.From.FullAddress;
    dlg.edtTo.Text := AMsg.ToList.EmailAddresses;
    dlg.edtSubject.Text := AMsg.Subject;
    dlg.memBody.Lines := AMsg.MessageText;

    dlg.ShowModal();
  finally
    dlg.Free();
  end;
end;

end.

File: CBuilder/ImapClient/MessageFrm.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "MessageFrm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TMessageForm *MessageForm;
//---------------------------------------------------------------------------
__fastcall TMessageForm::TMessageForm(TComponent* Owner)
        : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TMessageForm::ShowMessage(TclMailMessage *AMsg)
{
  TMessageForm *dlg = new TMessageForm(NULL);
  __try {
    dlg->btnAppend->Visible = false;
    dlg->btnCancel->Caption = "Close";

    dlg->edtFrom->ReadOnly = true;
    dlg->edtTo->ReadOnly = true;
    dlg->edtSubject->ReadOnly = true;
    dlg->memBody->ReadOnly = true;

    dlg->edtFrom->Color = clWindow;
    dlg->edtTo->Color = clWindow;
    dlg->edtSubject->Color = clWindow;
    dlg->memBody->Color = clWindow;

    dlg->edtFrom->Text = AMsg->From->FullAddress;
    dlg->edtTo->Text = AMsg->ToList->EmailAddresses;
    dlg->edtSubject->Text = AMsg->Subject;
    dlg->memBody->Lines = AMsg->MessageText;

    dlg->ShowModal();
  }
  __finally {
    delete dlg;
  }
}
//---------------------------------------------------------------------------

bool __fastcall TMessageForm::ComposeMessage(TclMailMessage *AMsg)
{
  bool res = false;
  TMessageForm *dlg = new TMessageForm(NULL);
  __try {
    dlg->btnAppend->Visible = true;
    dlg->btnCancel->Caption = "Cancel";

    dlg->edtFrom->ReadOnly = false;
    dlg->edtTo->ReadOnly = false;
    dlg->edtSubject->ReadOnly = false;
    dlg->memBody->ReadOnly = false;

    dlg->edtFrom->Text = "clevertester@company.mail";
    dlg->edtTo->Text = "clevertester@company.mail";
    dlg->edtSubject->Text = "Test Subject";
    dlg->memBody->Text = "Test Body";

    res = (dlg->ShowModal() == mrOk);

    if (res) {
      AMsg->BuildMessage(dlg->memBody->Lines->Text, "");
      AMsg->From->FullAddress = dlg->edtFrom->Text;
      AMsg->ToList->EmailAddresses = dlg->edtTo->Text;
      AMsg->Subject = dlg->edtSubject->Text;
    }
  }
  __finally {
    delete dlg;
  }
  return res;
}
